File: leetcode460/leetcode460_1.cpp
struct CacheNode {
    int key;
    int value;
    int freq;
    long tick;
    
    // Defines the order, smaller one will be evicted first
    bool operator <(const CacheNode& rhs) const {
        if (freq < rhs.freq)
            return true;
        if (freq == rhs.freq)
            return tick < rhs.tick;
        return false;
    }
};

class LFUCache {
public:
    LFUCache(int capacity):capacity_(capacity), tick_(0) {
        
    }
    
    int get(int key) {
        auto it = hashmap.find(key);
        if (it == hashmap.cend())
            return -1;
        int value = it->second.value;
        // update tick and freq
        touch(it->second);
        return value;
    }
    
    void put(int key, int value) {
        if (capacity_ == 0)
            return;
        
        auto it = hashmap.find(key);
        
        // already exists
        if (it != hashmap.cend()) {
            // key exists, update value and touch
            it->second.value = value;;
            touch(it->second);
            return;
        }
        
        // reach capacity
        if (hashmap.size() == capacity_) {
            // remove the first node in cache
            const CacheNode& node = *cache_.cbegin();
            hashmap.erase(node.key);
            cache_.erase(node);
        }
        
        // adding new node
        CacheNode node{key, value, 1, ++tick_};
        hashmap[node.key] = node;
        cache_.insert(node);
    }
private:
    void touch(CacheNode& node) {
        cache_.erase(node); // log(capacity)
        ++node.freq;
        node.tick = ++tick_;
        cache_.insert(node); // log(capacity)
    }
    
    long tick_;
    int capacity_;
    unordered_map<int, CacheNode> hashmap;
    set<CacheNode> cache_;
};

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache* obj = new LFUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

// Runtime: 212 ms, faster than 65.24% of C++ online submissions for LFU Cache.
// Memory Usage: 44.9 MB, less than 15.85% of C++ online submissions for LFU Cache.
// time complexity: O(log c)
// space complexity: O(c)

// http://zxi.mytechroad.com/blog/hashtable/leetcode-460-lfu-cache/
// // Author: Huahua
// // Running time: 99 ms
// struct CacheNode {
//     int key;
//     int value;
//     int freq;
//     long tick;
    
//     // Defines the order, smaller one will be evicted first
//     bool operator <(const CacheNode& rhs) const {
//         if (freq < rhs.freq) return true;
//         if (freq == rhs.freq) return tick < rhs.tick;
//         return false;
//     }
// };
// class LFUCache {
// public:
//     LFUCache(int capacity):capacity_(capacity), tick_(0) {}
    
//     int get(int key) {
//         auto it = m_.find(key);
//         if (it == m_.cend()) return -1;
//         int value = it->second.value;
//         touch(it->second);
//         return value;
//     }
    
//     void put(int key, int value) {
//         if (capacity_ == 0) return;
        
//         auto it = m_.find(key);
        
//         if (it != m_.cend()) {
//             // Key exists, update value and touch
//             it->second.value = value;            
//             touch(it->second);            
//             return;
//         }
        
//         if (m_.size() == capacity_) {
//             // Remove the first node in cache
//             const CacheNode& node = *cache_.cbegin();
//             m_.erase(node.key);
//             cache_.erase(node);
//         }
        
//         CacheNode node{key, value, 1, ++tick_};
//         m_[node.key] = node;
//         cache_.insert(node);
//     }
// private:
//     void touch(CacheNode& node) {
//         cache_.erase(node);     // log(capacity)        
//         ++node.freq;
//         node.tick = ++tick_;
//         cache_.insert(node);    // log(capacity)
//     }
    
//     long tick_;
//     int capacity_;
//     unordered_map<int, CacheNode> m_;
//     set<CacheNode> cache_;
    
// };

File: leetcode460/leetcode460_1.java
public class LFUCache {
    // min frequency in cache
    private int min;

    private final int capacity;
    // given key, get value
    private final HashMap<Integer, Integer> keyToVal;
    // given key, get frequency
    private final HashMap<Integer, Integer> keyToCount;
    // given frequency, get keys
    // a LinkedHashSet lets us iterate through the elements in the order in which they were inserted.
    // https://www.geeksforgeeks.org/linkedhashset-in-java-with-examples/
    private final HashMap<Integer, LinkedHashSet<Integer>> countToLRUKeys;
    
    public LFUCache(int capacity) {
        this.min = -1;
        this.capacity = capacity;
        this.keyToVal = new HashMap<>();
        this.keyToCount = new HashMap<>();
        this.countToLRUKeys = new HashMap<>();
    }
    
    // each time get key will update the frequency
    public int get(int key) {
        if (!keyToVal.containsKey(key)) return -1;
        
        int count = keyToCount.get(key);
        countToLRUKeys.get(count).remove(key); // remove key from current count (since we will inc count)
        if (count == min && countToLRUKeys.get(count).size() == 0) min++; // nothing in the current min bucket
        
        putCount(key, count + 1);
        return keyToVal.get(key);
    }
    
    public void put(int key, int value) {
        if (capacity <= 0) return;
        
        if (keyToVal.containsKey(key)) {
            keyToVal.put(key, value); // update key's value
            get(key); // update key's count
            return;
        } 
        
        if (keyToVal.size() >= capacity)
            evict(countToLRUKeys.get(min).iterator().next()); // evict LRU from this min count bucket
        
        min = 1;
        putCount(key, min); // adding new key and count
        keyToVal.put(key, value); // adding new key and value
    }
    
    // remove the key with least frequency
    private void evict(int key) {
        // remove from frequency to key
        countToLRUKeys.get(min).remove(key);
        // remove from key to value
        keyToVal.remove(key);
        // remove from key to frequency
        keyToCount.remove(key);
    }
    
    private void putCount(int key, int count) {
        keyToCount.put(key, count);
        countToLRUKeys.computeIfAbsent(count, ignore -> new LinkedHashSet<>());
        countToLRUKeys.get(count).add(key);
    }
}

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache obj = new LFUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

 
// Runtime: 89 ms, faster than 14.35% of Java online submissions for LFU Cache.
// Memory Usage: 122 MB, less than 31.24% of Java online submissions for LFU Cache.
// time complexity: O(1)
// space complexity: O(c)
// https://leetcode.com/problems/lfu-cache/discuss/94521/JAVA-O(1)-very-easy-solution-using-3-HashMaps-and-LinkedHashSet